hs_collection_resource/views.py

```diff
@@ -47,22 +54,54 @@ def update_collection(request, shortkey, *args, **kwargs):
                 if updated_contained_res_id == shortkey:
                     raise Exception("Can not add collection itself.")
 
-                # check authorization for all new resources being added to the collection
-                # the requesting user should at least have metadata view permission for each of the
-                # new resources to be added to the collection
-                if not collection_res_obj.\
-                        resources.filter(short_id=updated_contained_res_id).exists():
+            # current contained res
+            res_id_list_current_collection = \
+                [res.short_id for res in collection_res_obj.resources.all()]
+
+            # res to remove
+            res_id_list_remove = []
+            for res_id_remove in res_id_list_current_collection:
+                if res_id_remove not in updated_contained_res_id_list:
+                    res_id_list_remove.append(res_id_remove)
+                    res_obj_remove = get_resource_by_shortkey(res_id_remove)
+                    collection_res_obj.resources.remove(res_obj_remove)
+
+                    # change "Relation" metadata in collection
+                    value = relation_value_template.format(res_id_remove)
+                    add_or_remove_relation_metadata(add=False, target_res_obj=collection_res_obj,
+                                                    relation_type=hasPart, relation_value=value,
+                                                    set_res_modified=False)
+
+            # res to add
+            res_id_list_add = []
+            for res_id_add in updated_contained_res_id_list:
+                if res_id_add not in res_id_list_current_collection:
+                    res_id_list_add.append(res_id_add)
+                    # check authorization for all new resources being added to the collection
+                    # the requesting user should at least have metadata view permission for each of
+                    # the new resources to be added to the collection
                     res_to_add, _, _ \
-                        = authorize(request, updated_contained_res_id,
+                        = authorize(request, res_id_add,
                                     needed_permission=ACTION_TO_AUTHORIZE.VIEW_METADATA)
 
-            # remove all resources from the collection
-            collection_res_obj.resources.clear()
+                    is_shareable = res_to_add.raccess.shareable
+                    is_owner = res_to_add.raccess.owners.filter(pk=user.pk).exists()
+                    if not is_shareable and not is_owner:
+                            raise Exception('Only resource owner can add a resource '
```

Comment:
Should this message be: "Only resource owner can add a non-shareable resource to a collection"?